EXPERIMENT-6
-------------
AADITYA RAJ DIXIT
----------------
BATCH-49
--------------
SAP:500122014
------------------


In [10]:
# prompt: get a csv file from user as an input

from google.colab import files

uploaded = files.upload()

for fn in uploaded.keys():
  print('User uploaded file "{name}" with length {length} bytes'.format(
      name=fn, length=len(uploaded[fn])))


Saving electronic-card-transactions-august-2024-csv-tables.xlsx.csv to electronic-card-transactions-august-2024-csv-tables.xlsx (1).csv
User uploaded file "electronic-card-transactions-august-2024-csv-tables.xlsx (1).csv" with length 3757719 bytes


In [13]:


import pandas as pd
import numpy as np

# Assuming the uploaded file is a CSV, read it into a pandas DataFrame
try:
    df = pd.read_csv(list(uploaded.keys())[0])
except Exception as e:
    print("Error reading the file. Please ensure it's a valid CSV file.")
    print(e)
    # You may want to add more robust error handling here

# --- Missing Value Handling ---
# 1. Identify missing values
print("Missing values before handling:\n", df.isnull().sum())

# 2. Handle missing values (replace with mean, median, or mode)
for column in df.columns:
    if df[column].isnull().any():
        if pd.api.types.is_numeric_dtype(df[column]):
            # Use method without inplace to avoid FutureWarning
            df[column] = df[column].fillna(df[column].mean())
        else:
            df[column] = df[column].fillna(df[column].mode()[0])

# 3. Verify if missing values are handled
print("Missing values after handling:\n", df.isnull().sum())

# --- Outlier Handling ---
# 1. Identify outliers using IQR (Interquartile Range)
# Store bounds in a dictionary for each column
bounds = {}

for column in df.columns:
    if pd.api.types.is_numeric_dtype(df[column]):
        Q1 = df[column].quantile(0.25)
        Q3 = df[column].quantile(0.75)
        IQR = Q3 - Q1
        lower_bound = Q1 - 1.5 * IQR
        upper_bound = Q3 + 1.5 * IQR

        # Store the bounds for the column
        bounds[column] = (lower_bound, upper_bound)

        # Filter rows based on the bounds (remove outliers)
        df = df[(df[column] >= lower_bound) & (df[column] <= upper_bound)]

# 2. Handle outliers (cap the outliers at nearest valid value)
for column in df.columns:
    if pd.api.types.is_numeric_dtype(df[column]):
        lower_bound, upper_bound = bounds[column]
        # Use np.clip to cap values at the bounds
        df[column] = np.clip(df[column], lower_bound, upper_bound)

# Output data after handling outliers
print("Data after outlier handling:\n", df.head())


Missing values before handling:
 Series_reference        0
Period                  0
Data_value           1127
Suppressed          20691
STATUS                  0
UNITS                   0
Magnitude               0
Subject                 0
Group                   0
Series_title_1          0
Series_title_2          0
Series_title_3      16632
Series_title_4      20691
Series_title_5      20691
dtype: int64
Missing values after handling:
 Series_reference        0
Period                  0
Data_value              0
Suppressed          20691
STATUS                  0
UNITS                   0
Magnitude               0
Subject                 0
Group                   0
Series_title_1          0
Series_title_2          0
Series_title_3          0
Series_title_4      20691
Series_title_5      20691
dtype: int64
Data after outlier handling:
 Empty DataFrame
Columns: [Series_reference, Period, Data_value, Suppressed, STATUS, UNITS, Magnitude, Subject, Group, Series_title_1, Series_title_2, S

Missing values before handling:
--------------------------------
 Series_reference        0

Period                  0

Data_value           1127

Suppressed          20691

STATUS                  0

UNITS                   0

Magnitude               0

Subject                 0

Group                   0

Series_title_1          0

Series_title_2          0

Series_title_3      16632

Series_title_4      20691

Series_title_5      20691

dtype: int64

Missing values after handling:
----------------------------

 Series_reference        0

Period                  0

Data_value              0

Suppressed          20691

STATUS                  0

UNITS                   0

Magnitude               0

Subject                 0

Group                   0

Series_title_1          0

Series_title_2          0

Series_title_3          0

Series_title_4      20691

Series_title_5      20691

dtype: int64

Data after outlier handling:

 Empty DataFrame

Columns: [Series_reference, Period, Data_value, Suppressed, STATUS, UNITS, Magnitude, Subject, Group, Series_title_1, Series_title_2, Series_title_3, Series_title_4, Series_title_5]

Index: []